### PR TITLE
BUG: stats.ks_1samp: honour method argument for one-sided alternatives

### PR DESCRIPTION
**Pull request by m4cd4r4**
#### Reference issue

Related to gh-24737 (item 2 - `method` ignored for one-sided alternatives).

#### What does this implement/fix?

When `alternative='greater'` or `alternative='less'`, `ks_1samp` was using the exact one-sided distribution (`ksone.sf`) unconditionally, ignoring the `method` argument. The two-sided path already had a proper three-way branch on `mode` (exact/asymp/approx); the one-sided paths had none.

This PR adds the equivalent branch for one-sided alternatives:

- `method='exact'` (and `'auto'`, `'approx'`): uses `ksone.sf` as before
- `method='asymp'`: uses the Smirnov (1948) asymptotic formula `exp(-2 * n * d^2)`

A test is added to confirm that `method='exact'` and `method='asymp'` produce different p-values for one-sided alternatives (the pre-fix behaviour gave identical results).

The `xp.clip` call added to the one-sided returns mirrors the clip already applied to two-sided results, ensuring p-values stay in [0, 1].

#### Additional information

Item 1 of gh-24737 (invalid method string silently treated as `'approx'`) was fixed in gh-24784, which is now merged.

#### AI Generation Disclosure

Claude (Anthropic) was used to assist with research and drafting this fix. The asymptotic formula, code structure, and test were reviewed and verified by the contributor.

### CODE REVIEW

File: scipy/stats/tests/test_stats.py [hunk 1]
```diff
@@ -4718,6 +4718,24 @@ def cdf(x): return x / 10.
         xp_assert_equal(res.statistic_location, xp.asarray(ref_location))
         xp_assert_equal(res.statistic_sign, xp.asarray(ref_sign, dtype=xp.int8))
 
+    def test_method_honoured_for_one_sided(self):
+        # gh-24737: `method` was ignored for one-sided alternatives; verify
+        # that 'exact' and 'asymp' give different p-values.
+        rng = np.random.default_rng(594235924652)
+        x = rng.standard_normal(100)
+        for alt in ('greater', 'less'):
```
Comment (tylerjereddy) [inline scipy/stats/tests/test_stats.py]: May be sensible to hoist this up as a parametrization--that way if there is a failure in just one of the cases, we can better assess what might be going on.

File: scipy/stats/tests/test_stats.py [hunk 2]
```diff
@@ -4718,6 +4718,24 @@ def cdf(x): return x / 10.
...
+            res_exact = stats.ks_1samp(x, special.ndtr, alternative=alt,
+                                       method='exact')
+            res_asymp = stats.ks_1samp(x, special.ndtr, alternative=alt,
+                                       method='asymp')
```
Comment (tylerjereddy) [inline scipy/stats/tests/test_stats.py]: I think Matt's issue also mentions `kstest` for "point 2," so testing that would likely also be sensible

File: scipy/stats/tests/test_stats.py [hunk 3]
```diff
@@ -4718,6 +4718,24 @@ def cdf(x): return x / 10.
...
+            # statistic must be identical regardless of method
+            assert_equal(res_exact.statistic, res_asymp.statistic)
```
Comment (tylerjereddy) [inline scipy/stats/tests/test_stats.py]: Might be slightly better to use `assert_allclose()` since these are floats. Even if you really do expect them to be identical for now, it can help prevent flakes that might come in after algorithm changes, etc.

File: scipy/stats/tests/test_stats.py [hunk 4]
```diff
@@ -4718,6 +4718,24 @@ def cdf(x): return x / 10.
...
+            # p-values must differ (exact uses ksone.sf, asymp uses exp formula)
+            assert res_exact.pvalue != res_asymp.pvalue, (
```
Comment (tylerjereddy) [inline scipy/stats/tests/test_stats.py]: While this does indeed fail before the patch is applied, it may not be an ideal way to regression test that things are working properly. For example, if one of the methods suddenly starts returning a broken result, and the other doesn't, this would still make them unequal.

Could we possibly get expected outcomes for a few scenarios from ie., the R community? We can't look at their source because of licensing, but can record outputs from execution as expected values.

File: scipy/stats/tests/test_stats.py [hunk 5]
```diff
@@ -4718,6 +4718,24 @@ def cdf(x): return x / 10.
...
+                f"method='exact' and method='asymp' gave the same p-value "
+                f"for alternative='{alt}'; method is still being ignored"
```
Comment (tylerjereddy) [inline scipy/stats/tests/test_stats.py]: I think it is less common that we write custom error strings like this. Maybe not a big deal, but I believe we usually depend on the "native" error reporting.

**Comment by m4cd4r4** [comment]
Thanks for the thorough review @tylerjereddy - all 5 points addressed in the latest push:

1. **Parametrize** - hoisted alternatives into `@pytest.mark.parametrize`
2. **kstest coverage** - added `test_kstest_method_honoured_for_one_sided` for the wrapper
3. **assert_allclose** - switched throughout
4. **Reference values** - hardcoded expected p-values from `ksone.sf` (exact) and `exp(-2*n*d**2)` (asymptotic) so each method is validated independently. I don't have R installed locally to cross-check, but these follow directly from the known formulas - happy to verify against R output if needed.
5. **Custom error strings** - removed

Also rebased onto latest main.

**Comment by m4cd4r4** [comment]
Thanks for the review @tylerjereddy! All five points addressed in the latest two commits:

1. **Parametrize tests** - both `test_method_honoured_for_one_sided` and the new `test_kstest_method_honoured_for_one_sided` now use `@pytest.mark.parametrize` over `(alternative, ref_exact, ref_asymp)` triples instead of looping internally.

2. **kstest coverage** - added `test_kstest_method_honoured_for_one_sided` exercising the `kstest` wrapper for both alternatives and both methods, mirroring point 2 of gh-24737.

3. **assert_allclose** - all float comparisons now use `assert_allclose` rather than `assert_equal` or bare `assert`.

4. **External reference values** - the hardcoded p-values are now from R 4.3.3 `ks.test()`, not from scipy itself, so they validate against an independent implementation. Reproduction recipe is documented in a comment above the parametrize block:
    ```r
    x <- scan("x.csv")  # samples from default_rng(594235924652)
    ks.test(x, "pnorm", alternative="greater", exact=TRUE)$p.value
    # 0.81398628551406649
    ```
    The R values agree with scipy to ~14 digits.

5. **Custom error strings** - dropped, relying on the native pytest assertion output now.

Happy to make further adjustments.

**Comment by m4cd4r4** [comment]
Hi @tylerjereddy, wanted to check if you had a chance to review the changes I made on April 8 addressing your points. Let me know if anything else needs adjustment.

**Comment by ev-br** [comment]
CI errors look related:

```
=========================== short test summary info ============================
FAILED scipy/stats/tests/test_marray.py::test_ks_1samp[torch-0-less-asymp-ks_1samp] - DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
FAILED scipy/stats/tests/test_marray.py::test_ks_1samp[torch-0-less-asymp-kstest] - DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
FAILED scipy/stats/tests/test_marray.py::test_ks_1samp[torch-0-greater-asymp-ks_1samp] - DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
FAILED scipy/stats/tests/test_marray.py::test_ks_1samp[torch-0-greater-asymp-kstest] - DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
FAILED scipy/stats/tests/test_marray.py::test_ks_1samp[torch-1-less-asymp-ks_1samp] - DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
FAILED scipy/stats/tests/test_marray.py::test_ks_1samp[torch-1-less-asymp-kstest] - DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
FAILED scipy/stats/tests/test_marray.py::test_ks_1samp[torch-1-greater-asymp-ks_1samp] - DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
FAILED scipy/stats/tests/test_marray.py::test_ks_1samp[torch-1-greater-asymp-kstest] - DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
FAILED scipy/stats/tests/test_marray.py::test_ks_1samp[torch-None-less-asymp-ks_1samp] - DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
FAILED scipy/stats/tests/test_marray.py::test_ks_1samp[torch-None-less-asymp-kstest] - DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
FAILED scipy/stats/tests/test_marray.py::test_ks_1samp[torch-None-greater-asymp-ks_1samp] - DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
FAILED scipy/stats/tests/test_marray.py::test_ks_1samp[torch-None-greater-asymp-kstest] - DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
= 12 failed, 19105 passed, 3707 skipped, 73033 deselected, 52 xfailed, 2 xpassed in 123.11s (0:02:03) =
```

**Comment by m4cd4r4** [comment]
Thanks @ev-br, you're right that those are from this PR. Diagnosis and fix pushed.

The one-sided `asymp` branch I added computed `exp(-2 n D^2)` with `np.exp` inside a small helper run through `_masked_apply`. `_masked_apply` unwraps `MArray` arguments to their backing array before calling the helper, so on a non-NumPy backend (torch here) `np.exp` was handed a torch tensor and dispatched through `__array_wrap__`, raising the NumPy 2.0 `must accept context and return_scalar arguments` DeprecationWarning. Under `-W error` that failed the 12 `test_marray.py::test_ks_1samp[torch-*-asymp-*]` cases (both alternatives x 3 axes x the two wrappers).

The asymptotic p-value is plain elementwise math, so the fix just computes it with `xp.exp` directly on the (masked) statistic, keeping it in the array's own namespace. The mask propagates through the elementwise ops, the same way the existing two-sided `asymp` branch already does `D * N**0.5`. Values are unchanged.

I couldn't run the full `test_marray` suite locally (no built scipy on this box), but I reproduced the exact mechanism against numpy 2.2.3 + torch 2.6: `np.exp` on a torch tensor raises that DeprecationWarning (and errors under `-W error`), `xp.exp`/`torch.exp` does not, and the two give bit-identical results. CI should confirm across the other backends.

Disclosure: I used an AI assistant (Claude) to help trace the `__array_wrap__` dispatch and draft the fix; I reviewed and verified it.
